auth/src/source.rs

```diff
@@ -0,0 +1,474 @@
+        let token_response: TokenResponse = res
+            .json()
+            .await
+            .map_err(|_| Error::Other("unable to decode response".into()))?;
+        Ok(AccessToken {
+            value: token_response.access_token,
+            expires: Some(Utc::now() + Duration::seconds(token_response.expires_in)),
+        })
+    }
+}
+
+#[async_trait]
+impl Source for UserSource {
+    async fn token(&self) -> Result<AccessToken> {
+        self._access_token().await
+    }
+}
+
+/// The request body for talking to `https://oauth2.googleapis.com/token`.
+#[derive(Serialize)]
+struct UserTokenRequest<'a> {
+    grant_type: &'a str,
+    refresh_token: &'a str,
+    client_id: &'a str,
+    client_secret: &'a str,
+}
+
+pub struct ComputeSourceBuilder {
+    scopes: Vec<String>,
+}
+
+impl ComputeSourceBuilder {
+    pub fn scopes(mut self, scopes: Vec<String>) -> Self {
+        self.scopes.extend(scopes);
+        self
+    }
+
+    pub fn build(self) -> Result<ComputeSource> {
+        Ok(ComputeSource {
+            scopes: self.scopes,
+        })
+    }
+}
+
+/// A [Source] derived from the Google Cloud metadata service.
+pub struct ComputeSource {
+    scopes: Vec<String>,
+}
+
+impl ComputeSource {
+    pub fn builder() -> ComputeSourceBuilder {
+        ComputeSourceBuilder { scopes: Vec::new() }
+    }
+
+    async fn _access_token(&self) -> Result<AccessToken> {
+        let token = metadata::access_token(None, self.scopes.clone())
+            .await
+            .map_err(|e| Error::Other(e.to_string()))?;
+        Ok(AccessToken {
+            value: token.access_token,
+            expires: Some(Utc::now() + Duration::seconds(token.expires_in)),
+        })
+    }
+}
+
+#[async_trait]
+impl Source for ComputeSource {
+    async fn token(&self) -> Result<AccessToken> {
+        self._access_token().await
+    }
+}
+
+/// A noop source used for default credentials. It will never produce tokens.
+pub(crate) struct NoOpSource {}
+
+#[async_trait]
+impl Source for NoOpSource {
+    async fn token(&self) -> Result<AccessToken> {
+        Err(Error::Other(
+            "use Credential.find_default to find a credential from the env".into(),
+        ))
+    }
+}
+
+/// This type is meant to wrap another [Source] and keep returning the same [AccessToken]
+// as long as it is valid.
+pub(crate) struct RefresherSource {
+    pub(crate) current_token: Arc<Mutex<AccessToken>>,
+    pub(crate) source: Box<dyn Source + Send + Sync>,
+}
+
+impl Default for RefresherSource {
+    fn default() -> Self {
+        Self {
+            current_token: Arc::new(Mutex::new(AccessToken {
+                value: String::new(),
+                expires: None,
+            })),
+            source: Box::new(NoOpSource {}),
+        }
+    }
+}
+
+#[async_trait]
+impl Source for RefresherSource {
+    async fn token(&self) -> Result<AccessToken> {
+        let mut cur_token = self.current_token.lock().await;
+        if cur_token.needs_refresh() {
```

Comment:
The condition looks inverted.  Should this be `if !cure_token.needs_refresh() {`?

Reply:
I expect either that I'm misinterpreting the logic or a test to fail.

Reply:
When I re-named I implied the wrong thing. Updated name and tests